### PR TITLE
[Proposal] Types refactor for clean up of keeper

### PR DESCRIPTION
**Pull request by jackzampolin**
The following is a PR to propose the shape and scope of the general refactor of the go side of peggy. The major change here will be to put both `EthereumClaim`s and `Confirms` behind interfaces and simplify the external and internal interfaces to use these generalized types. This should result in significantly more readable, maintainable, and extensible code.

cc @fedekunze @jkilpatr @jtremblek

### CODE REVIEW

File: module/proto/peggy/v1/attestation.proto [hunk 1]
```diff
@@ -12,14 +10,24 @@ option go_package = "github.com/althea-net/peggy/module/x/peggy/types";
 // be handled
 enum ClaimType {
   option (gogoproto.goproto_enum_prefix) = false;
-
-  CLAIM_TYPE_UNKNOWN  = 0 [(gogoproto.enumvalue_customname) = "CLAIM_TYPE_UNKNOWN" ];
-  CLAIM_TYPE_DEPOSIT  = 1 [(gogoproto.enumvalue_customname) = "CLAIM_TYPE_DEPOSIT" ];
-  CLAIM_TYPE_WITHDRAW = 2 [(gogoproto.enumvalue_customname) = "CLAIM_TYPE_WITHDRAW"];
-  CLAIM_TYPE_ERC20_DEPLOYED = 3 [(gogoproto.enumvalue_customname) = "CLAIM_TYPE_ERC20_DEPLOYED"];
+  
+  CLAIM_TYPE_UNKNOWN             = 0 [(gogoproto.enumvalue_customname) = "CLAIM_TYPE_UNKNOWN" ];
```
Comment (jkilpatr) [inline module/proto/peggy/v1/attestation.proto:14]: I think we should remove claim type unknown. I mean if we don't know what it is we should panic and or refuse the input.

File: module/proto/peggy/v1/attestation.proto [hunk 2]
```diff
@@ -12,14 +10,24 @@ option go_package = "github.com/althea-net/peggy/module/x/peggy/types";
...
+  CLAIM_TYPE_WITHDRAW            = 2 [(gogoproto.enumvalue_customname) = "CLAIM_TYPE_WITHDRAW"];
+  CLAIM_TYPE_ERC20_DEPLOYED      = 3 [(gogoproto.enumvalue_customname) = "CLAIM_TYPE_ERC20_DEPLOYED"];
   CLAIM_TYPE_LOGIC_CALL_EXECUTED = 4 [(gogoproto.enumvalue_customname) = "CLAIM_TYPE_LOGIC_CALL_EXECUTED"];
 }
 
+// ConfirmType is the cosmos type of a confirmation of a given object from the eth signers
+enum ConfirmType {
+  option (gogoproto.goproto_enum_prefix) = false;
+
+  CONFIRM_TYPE_UNKNOWN  = 1 [(gogoproto.enumvalue_customname) = "CONFIRM_TYPE_UNKNOWN"];
```
Comment (jkilpatr) [inline module/proto/peggy/v1/attestation.proto:25]: repeat of the above.

**Review by jkilpatr — APPROVED** [review]
I'm happy with the way this proto file is laid out as a target for our improvements. We agree on the end goal.

What I don't think is a good idea is trying to tackle all the proto changes at once. We can break this down into several parts, and each should be done independently.

1) I need to get proto-tx's in the orchestrator, ETA is about a week.
2) The keeper refactor is generally independent of how messages are done any one of us can go and clean that up while (1) is being worked on
3) making a claims interface and migrating all software to there
4) making a confirm interface and migrating all software to there
5) updating endpoints to normalize naming.

3 and 4 depend on 1

2, 5 have no dependencies and in fact 5 could be split up into one pull for each endpoint. But we should group them in ways that make sense.

I think this is a reasonable way to scope the work, this is really important since we need to keep testing both Althea chain and Somm with development of both moving independently of this refactor.

File: module/proto/peggy/v1/attestation.proto (hunk 1 above)
Comment (marbar3778) [reply module/proto/peggy/v1/attestation.proto:14]: What is the default case? If we set 0 to CLAIM_TYPE_DEPOSIT then if the user does not set it we will assume its deposit. Setting Unknown is a recommended practice for protobuf and can help us in rejecting messages.

File: module/proto/peggy/v1/attestation.proto (hunk 2 above)
Comment (marbar3778) [reply module/proto/peggy/v1/attestation.proto:25]: This should be 0 unless unknown here is an actual type that is used?

File: module/proto/peggy/v1/attestation.proto (hunk 1 above)
Comment (jkilpatr) [reply module/proto/peggy/v1/attestation.proto:14]: oh wait this is a way we deal with go's really dumb and dangerous default behavior? I suppose that's acceptable

Comment (marbar3778) [reply module/proto/peggy/v1/attestation.proto:14]: Enums in proto should start at 0 in proto, 0 is generally used as the default value. It is also common practice in go.

File: module/proto/peggy/v1/msgs.proto [hunk 3]
```diff
@@ -110,114 +47,27 @@ message MsgSendToEth {
...
-message MsgDepositClaim {
-  uint64 event_nonce     = 1;
-  uint64 block_height    = 2;
-  string token_contract  = 3;
-  string amount          = 4 [
-    (gogoproto.customtype) = "github.com/cosmos/cosmos-sdk/types.Int", 
-    (gogoproto.nullable) = false
+message MsgSubmitClaim {
+  google.protobuf.Any claim = 1 [
+      (cosmos_proto.accepts_interface) = "EthereumClaim"
```
Comment (marbar3778) [inline module/proto/peggy/v1/msgs.proto:59]: Is the reason for using any here to allow other people to plug in their own claim types? If we want to allow users to only use the provided claim types we should use a oneof here. 

Its better to opinionated in a few things to reduce the complexity of such a system.

Comment (jackzampolin) [reply module/proto/peggy/v1/msgs.proto:59]: Thats a great point! And sounds like we should use a OneOf

**Comment by jackzampolin** [comment]
Closing this in favor of upstream issue.

**Pull request closed without merging**